gnocchi/notifier/redis.py

```diff
@@ -0,0 +1,51 @@
+# -*- encoding: utf-8 -*-
+#
+# Copyright © 2017 Red Hat
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License. You may obtain
+# a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+import struct
+
+from gnocchi.common import redis
+from gnocchi import notifier
+
+
+class RedisNotifier(notifier.Notifier):
+
+    _SERIALIZATION_FORMAT_LEN = struct.calcsize("<H")
+    DEFAULT_CHANNEL = b"gnocchi-measures"
+
+    def __init__(self, conf):
+        super(RedisNotifier, self).__init__(conf)
+        self._client, options = redis.get_client(conf.url)
+        self.channel = options.get('channel',
+                                   [self.DEFAULT_CHANNEL])[-1].decode()
+
+    def __str__(self):
+        return "%s: %s channel %s" % (
+            self.__class__.__name__, self._client, self.channel
+        )
+
+    def iter_on_sacks_to_process(self):
+        p = self._client.pubsub()
+        p.subscribe(self.channel)
+        for message in p.listen():
+            if (message['type'] == 'message' and
+               message['channel'] == self.NOTIFY_CHANNEL_NAME):
+                data = message['data']
+                nb_sack = data / self._SERIALIZATION_FORMAT_LEN
+                for sack in struct.unpack("<" + "H" * nb_sack):
+                    yield sack
+
+    def notify_new_measures_for_sacks(self, sacks):
+        self._client.publish(self.channel,
+                             struct.pack("<" + "H" * len(sacks), *sacks))
```

Comment:
Haha, it took you time to get on board, but here you come! :)

So yeah that is another solution to have one channel per sack, so that'd avoid spamming all the metricd and then filter. The only downsides that I see is that 1. it is trickier to implement because the sack list changes each time a worker join/leaves the tooz group and 2. it can therefore miss a little bit more notification than the current solution.

That means the subscription patterns must be regularly checked and updated as soon as tooz group member list changes.

I am currently not sure it is worth it – or not.

As for the API, `NOTIFY` should not slow anything down AFAIU. There might be also a way to leverage https://redis.io/topics/notifications which would avoid doing the `NOTIFY` call at all. I did not pursue this yet as 1. it's an optimization 2. at first glance it was not obvious it is something that you can control from the Redis client.